changelogs/unreleased/gh-6754-term-bump-before-promote.md

```diff
@@ -0,0 +1,4 @@
+## bugfix/raft
```

Comment:
I managed to create a stable reproducer, see diff below and a test:
<details><summary>diff</summary>

```Diff
diff --git a/src/box/lua/info.c b/src/box/lua/info.c
index 8d7bb1d4a..76cc50ada 100644
--- a/src/box/lua/info.c
+++ b/src/box/lua/info.c
@@ -644,6 +644,8 @@ lbox_info_synchro(struct lua_State *L)
 	lua_setfield(L, -2, "owner");
 	lua_pushboolean(L, latch_is_locked(&queue->promote_latch));
 	lua_setfield(L, -2, "busy");
+	luaL_pushuint64(L, queue->promote_greatest_term);
+	lua_setfield(L, -2, "term");
 	lua_setfield(L, -2, "queue");
 
 	return 1;
diff --git a/src/box/relay.cc b/src/box/relay.cc
index 1c588f70a..8281b7656 100644
--- a/src/box/relay.cc
+++ b/src/box/relay.cc
@@ -141,8 +141,14 @@ struct relay {
 	 */
 	struct vclock local_vclock_at_subscribe;
 
-	/** Relay endpoint */
-	struct cbus_endpoint endpoint;
+	/** Endpoint to receive messages from WAL. */
+	struct cbus_endpoint wal_endpoint;
+	/**
+	 * Endpoint to receive messages from TX. Having the 2 endpoints
+	 * separated helps to synchronize the data coming from TX and WAL. Such
+	 * as term bumps from TX with PROMOTE rows from WAL.
+	 */
+	struct cbus_endpoint tx_endpoint;
 	/** A pipe from 'relay' thread to 'tx' */
 	struct cpipe tx_pipe;
 	/** A pipe from 'tx' thread to 'relay' */
@@ -811,7 +817,7 @@ relay_send_is_raft_enabled(struct relay *relay,
 	 * fiber.
 	 */
 	while (!msg->is_finished) {
-		cbus_process(&relay->endpoint);
+		cbus_process(&relay->tx_endpoint);
 		if (msg->is_finished)
 			break;
 		fiber_yield();
@@ -831,12 +837,16 @@ relay_subscribe_f(va_list ap)
 	coio_enable();
 	relay_set_cord_name(relay->io->fd);
 
-	/* Create cpipe to tx for propagating vclock. */
-	cbus_endpoint_create(&relay->endpoint, tt_sprintf("relay_%p", relay),
+	cbus_endpoint_create(&relay->tx_endpoint,
+			     tt_sprintf("relay_tx_%p", relay),
 			     fiber_schedule_cb, fiber());
-	cbus_pair("tx", relay->endpoint.name, &relay->tx_pipe,
+	cbus_pair("tx", relay->tx_endpoint.name, &relay->tx_pipe,
 		  &relay->relay_pipe, NULL, NULL, cbus_process);
 
+	cbus_endpoint_create(&relay->wal_endpoint,
+			     tt_sprintf("relay_wal_%p", relay),
+			     fiber_schedule_cb, fiber());
+
 	struct relay_is_raft_enabled_msg raft_enabler;
 	if (!relay->replica->anon && relay->version_id >= version_id(2, 6, 0))
 		relay_send_is_raft_enabled(relay, &raft_enabler, true);
@@ -852,7 +862,7 @@ relay_subscribe_f(va_list ap)
 		trigger_add(&relay->r->on_close_log, &on_close_log);
 
 	/* Setup WAL watcher for sending new rows to the replica. */
-	wal_set_watcher(&relay->wal_watcher, relay->endpoint.name,
+	wal_set_watcher(&relay->wal_watcher, relay->wal_endpoint.name,
 			relay_process_wal_event, cbus_process);
 
 	/* Start fiber for receiving replica acks. */
@@ -883,13 +893,15 @@ relay_subscribe_f(va_list ap)
 
 		fiber_cond_wait_deadline(&relay->reader_cond,
 					 relay->last_row_time + timeout);
-
 		/*
 		 * The fiber can be woken by IO cancel, by a timeout of
 		 * status messaging or by an acknowledge to status message.
 		 * Handle cbus messages first.
 		 */
-		cbus_process(&relay->endpoint);
+		inj = errinj(ERRINJ_RELAY_FROM_TX_DELAY, ERRINJ_BOOL);
+		if (inj == NULL || !inj->bparam)
+			cbus_process(&relay->tx_endpoint);
+		cbus_process(&relay->wal_endpoint);
 		/* Check for a heartbeat timeout. */
 		if (ev_monotonic_now(loop()) - relay->last_row_time > timeout)
 			relay_send_heartbeat(relay);
@@ -939,7 +951,8 @@ relay_subscribe_f(va_list ap)
 	/* Destroy cpipe to tx. */
 	cbus_unpair(&relay->tx_pipe, &relay->relay_pipe,
 		    NULL, NULL, cbus_process);
-	cbus_endpoint_destroy(&relay->endpoint, cbus_process);
+	cbus_endpoint_destroy(&relay->wal_endpoint, cbus_process);
+	cbus_endpoint_destroy(&relay->tx_endpoint, cbus_process);
 
 	relay_exit(relay);
 
diff --git a/src/lib/core/errinj.h b/src/lib/core/errinj.h
index 2c8edfa58..7b71bf375 100644
--- a/src/lib/core/errinj.h
+++ b/src/lib/core/errinj.h
@@ -169,6 +169,7 @@ struct errinj {
 	_(ERRINJ_XLOG_META, ERRINJ_BOOL, {.bparam = false}) \
 	_(ERRINJ_XLOG_READ, ERRINJ_INT, {.iparam = -1}) \
 	_(ERRINJ_XLOG_RENAME_DELAY, ERRINJ_BOOL, {.bparam = false}) \
+	_(ERRINJ_RELAY_FROM_TX_DELAY, ERRINJ_BOOL, {.bparam = false}) \
 
 ENUM0(errinj_id, ERRINJ_LIST);
 extern struct errinj errinjs[];
```
</details>

<details><summary>test</summary>

```Lua
--
-- Instance 1
--
-- Step 1
--
box.cfg{
    listen = 3313,
    replication = {3313, 3314},
    replication_synchro_quorum = 1,
    replication_timeout = 0.1,
    election_mode = 'manual',
}
box.schema.user.grant('guest', 'super')

-- Step 4
--
box.error.injection.set("ERRINJ_RELAY_FROM_TX_DELAY", true)
box.ctl.promote()
box.info.election.term      -- = 4
box.info.synchro.queue.term -- = 4

-- Step 6
--
box.error.injection.set("ERRINJ_RELAY_FROM_TX_DELAY", false)


--
-- Instance 2
--
-- Step 2
--
box.cfg{
    listen = 3314,
    replication = {3313, 3314},
    read_only = true,
    replication_synchro_quorum = 1,
    replication_timeout = 0.1,
    election_mode = 'manual',
}

-- Step 3
--
box.ctl.promote()

-- Step 5
--
box.info.election.term      -- = 3
box.info.synchro.queue.term -- = 4

-- Step 7
--
box.info.election.term      -- = 4
box.info.synchro.queue.term -- = 4
```
</details>

In the test I do an error injection to simulate too long delivery of TX messages. So as the relay reads WAL messages faster than gets TX messages.

The diff also allows you to implement the proposal about handling the order issue entirely inside of relay. Because now TX and WAL messages are delivered separately. When you meet `PROMOTE` with a too big term, you can spin in `cbus_process(&tx_endpoint)`. Dispatch of `wal_endpoint` won't grab TX messages alongside like it happened when you tried to implement it first time.

It is also worth mentioning that in this test we indeed can observe synchro term > raft term, but I couldn't make any real bug out of it. After you unblock the injection, the raft term is delivered just fine, not ignored.